serialization/serialization.go

```diff
@@ -0,0 +1,291 @@
+			switch f := m.Fields[i].Type.(type) {
+			case FieldIntFixed:
+				f.Value, err = decodeFixed(r, f.Length)
+				if err != nil {
+					return err
+				}
+				m.Fields[i].Type = f
+			case FieldIntVar:
+				f.Value, err = decodeVar(r, f.Unsigned)
+				if err != nil {
+					return err
+				}
+				m.Fields[i].Type = f
+			case FieldString:
+				f.Value, err = decodeString(r)
+				if err != nil {
+					return err
+				}
+				m.Fields[i].Type = f
+			default:
+				return fmt.Errorf("unsupported field type: %T", m.Fields[i].Type)
+			}
+		}
+
+	default:
+		return fmt.Errorf("unsupported type: %T", v)
+	}
+	return nil
+}
+
+func decodeString(r io.Reader) (string, error) {
+	firstByte := make([]byte, 1)
+	_, err := r.Read(firstByte)
+	if err != nil {
+		return "", err
+	}
+	strBytes := make([]byte, firstByte[0]/2)
+	n, err := r.Read(strBytes)
+	if err != nil {
+		return "", err
+	}
+	if n != int(firstByte[0]/2) {
+		return "", fmt.Errorf("only read %d bytes, expected %d", n, firstByte[0]/2)
+	}
+	return string(strBytes), nil
+}
+
+func decodeFixed(r io.Reader, len int) ([]byte, error) {
+	var b bytes.Buffer
+
+	tmpInt := make([]byte, 1)
+	for {
+		_, err := r.Read(tmpInt)
+		if err != nil {
+			return nil, err
+		}
+		if tmpInt[0]%2 == 0 {
+			b.WriteByte(tmpInt[0] / 2)
+		} else {
+			tmpInt2 := make([]byte, 1)
+			_, err := r.Read(tmpInt2)
+			if err != nil {
+				return nil, err
+			}
+			switch tmpInt2[0] {
+			case 0x2:
+				b.WriteByte((tmpInt[0] >> 2) + 0x80)
+			case 0x3:
+				b.WriteByte((tmpInt[0] >> 2) + 0xc0)
+			default:
+				return nil, fmt.Errorf("unknown decoding for %v", tmpInt2[0])
+			}
+		}
+		if b.Len() == len {
+			break
+		}
+	}
+	return b.Bytes(), nil
+}
+
+func decodeVar(r io.ReadSeeker, unsigned bool) (uint64, error) {
+	firstByte := make([]byte, 1)
+	_, err := r.Read(firstByte)
+	if err != nil {
+		return 0, err
+	}
+	tb := trailingOneBitCount(firstByte[0])
+	_, err = r.Seek(-1, io.SeekCurrent)
+	if err != nil {
+		return 0, err
+	}
+	fieldBytes := make([]byte, tb+1)
+	n, err := r.Read(fieldBytes)
+	if err != nil {
+		return 0, err
+	}
+	if n != tb+1{
+		return 0, fmt.Errorf("only read %d bytes, expected %d", n, tb+1)
+	}
+	var tNum uint64
+	switch len(fieldBytes) {
+	case 1:
+		tNum = uint64(fieldBytes[0])
+	case 2:
+		tNum = uint64(binary.LittleEndian.Uint16(fieldBytes))
+	case 3:
+		tNum = uint64(binary.LittleEndian.Uint32(
+			slices.Concat(fieldBytes, []byte{0x0})))
+	case 4:
+		tNum = uint64(binary.LittleEndian.Uint32(fieldBytes))
+	case 5:
+		tNum = binary.LittleEndian.Uint64(
+			slices.Concat(fieldBytes, []byte{0x0, 0x0, 0x0}))
+	case 6:
+		tNum = binary.LittleEndian.Uint64(
+			slices.Concat(fieldBytes, []byte{0x0, 0x0}))
+	case 7:
+		tNum = binary.LittleEndian.Uint64(
+			slices.Concat(fieldBytes, []byte{0x0}))
+	case 8:
+		tNum = binary.LittleEndian.Uint64(fieldBytes)
+	}
+	if unsigned {
+		return (tNum >> (tb + 2) * 2)+1, nil
+	}
+	return tNum >> (tb + 2), nil
+}
+
+func trailingOneBitCount(b byte) (count int) {
```

Comment:
we can revert the bits and use https://pkg.go.dev/math/bits#TrailingZeros8

Reply:
I assume you meant reverse instead of revert?

Reply:
Bitwise not, `^b`